ui/src/app/shared/services/workflows-service.ts

```diff
@@ -118,27 +120,66 @@ export class WorkflowsService {
             .then(res => res.body as Workflow);
     }
 
+    public getContainerLogsFromCluster(workflow: Workflow, nodeId: string, container: string): Observable<string> {
+        const podLogsURL = `api/v1/workflows/${workflow.metadata.namespace}/${workflow.metadata.name}/${nodeId}/log?logOptions.container=${container}&logOptions.follow=true`;
+        return requests
+            .loadEventSource(podLogsURL)
+            .map(line => JSON.parse(line).result.content)
+            .filter(isString)
+            .catch(() => {
+                // When an error occurs on an observable, RxJS is hard-coded to unsubscribe from the stream.  In the case
+                // that the connection to the server was interrupted while the node is still pending or running, this is not
+                // correct since we actually want the EventSource to re-connect and continue streaming logs.  In the event
+                // that the pod has completed, then we want to allow the unsubscribe to happen since no additional logs exist.
+                return Observable.fromPromise(this.isWorkflowNodePendingOrRunning(workflow, nodeId)).switchMap(isPendingOrRunning => {
+                    if (isPendingOrRunning) {
+                        return this.getContainerLogsFromCluster(workflow, nodeId, container);
+                    }
+
+                    // If our workflow is completed, then simply complete the Observable since nothing else
+                    // should be omitted
+                    return Observable.empty();
+                });
+            });
+    }
+
+    public async isWorkflowNodePendingOrRunning(workflow: Workflow, nodeId: string) {
+        if (this.isNodePendingOrRunning(workflow.status.nodes[nodeId])) {
+            return false;
+        }
+
+        // TODO: What happens if we're offline when we make this request?  We should figure out what errors we should
+        // retry on, and what errors we should allow to occur, then transform this to an Observable that implements
+        // all of that
+        const updatedWorkflow = await this.get(workflow.metadata.namespace, workflow.metadata.name);
+
+        return this.isNodePendingOrRunning(updatedWorkflow.status.nodes[nodeId]);
+    }
+
+    public getContainerLogsFromArtifact(workflow: Workflow, nodeId: string, container: string, archived: boolean) {
+        return Observable.of(this.hasArtifactLogs(workflow, nodeId, container))
+            .switchMap(hasArtifactLogs => {
+                if (!hasArtifactLogs) {
+                    throw new Error('no artifact logs are available');
+                }
+
+                return Observable.fromPromise(requests.get(this.getArtifactLogsUrl(workflow, nodeId, container, archived)));
```

Comment:
One notable problem here is that when you're running a local instance of the UI via `argo server`, this actually hangs for ~30s if your artifact repository is not available externally since you can't resolve the hostname if it's a `.cluster.local` address, so that is not a good user experience.

Reply:
For the non-archived case, I swapped this back to loading from the cluster first since the UX was impacted too much by this change imo.

Reply:
Note to self: test pod deletion and ensure that the fallback to log archives works correctly.  if it does, this will also correct https://github.com/argoproj/argo/issues/3953